src/fetchcode/package.py

```diff
@@ -327,3 +333,775 @@ def get_rubygems_data_from_purl(purl):
         download_url=download_url,
         **purl.to_dict(),
     )
+
+
+@router.route("pkg:gnu/.*")
+def get_gnu_data_from_purl(purl):
+    """Generate `Package` object from the `purl` string of gnu type"""
+    purl = PackageURL.from_string(purl)
+    source_archive_url = f"https://ftp.gnu.org/pub/gnu/{purl.name}/"
+    regex = r"^({}-)([\w.-]*)(.tar.gz)$".format(purl.name)
+
+    yield from extract_packages_from_listing(purl, source_archive_url, regex, [])
+
+
+@dataclasses.dataclass
+class DirectoryListedSource:
+    source_url: str = dataclasses.field(
+        default="", metadata={"description": "URL of the directory listing source"}
+    )
+    is_nested: bool = dataclasses.field(
+        default=False,
+        metadata={
+            "description": "Flag indicating whether the archives are nested within another directory"
+        },
+    )
+    source_archive_regex: str = dataclasses.field(
+        default="",
+        metadata={
+            "description": "Regular expression pattern to match files in the directory listing."
+        },
+    )
+    ignored_files_and_dir: List[str] = dataclasses.field(
+        default_factory=list,
+        metadata={
+            "description": "List of files and directories to ignore in the directory listing."
+        },
+    )
+
+    @classmethod
+    def get_package_info(cls, package_url):
+        if cls.is_nested:
+            yield from extract_package_from_nested_listing(
+                package_url,
+                cls.source_url,
+                cls.source_archive_regex,
+                cls.ignored_files_and_dir,
+            )
+
+        else:
+            yield from extract_packages_from_listing(
+                package_url,
+                cls.source_url,
+                cls.source_archive_regex,
+                cls.ignored_files_and_dir,
+            )
+
+
+class IpkgDirectoryListedSource(DirectoryListedSource):
+    source_url = "https://web.archive.org/web/20090326020239/http://handhelds.org/download/packages/ipkg/"
+    is_nested = False
+    source_archive_regex = r"^(ipkg[-_])([\w.-]*)(_arm.ipk|.tar.gz)$"
+    ignored_files_and_dir = []
+
+    @classmethod
+    def get_package_info(cls, package_url):
+        # Since there will be no new releases of ipkg, it's better to
+        # store them in a dictionary rather than fetching them every time.
+        archives = {
```

Comment:
I don't think having it hardcoded here makes sense, it can be stored in a different file

Reply:
Moved the release info to `src/fetchcode/ipkg_release_info.py` file.